src/NuGet.Core/NuGet.Protocol/Converters/IgnoreCaseStringKeyDictionaryConverter.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using NuGet.Protocol.Model;
+
+namespace NuGet.Protocol.Converters
+{
+    internal class VulnerabilityFileDataConverter : JsonConverter
+    {
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(VulnerabilityFileData);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer)
+        {
+            Dictionary<string, IReadOnlyList<VulnerabilityInfo>> result = new(StringComparer.InvariantCultureIgnoreCase);
+
+            if (!reader.Read())
+            {
+                throw new JsonSerializationException("Unexpected end of data");
+            }
+
+            bool finished = false;
+            do
+            {
+                switch (reader.TokenType)
+                {
+                    case JsonToken.EndObject:
+                        finished = true;
+                        break;
+
+                    case JsonToken.PropertyName:
+                        {
+                            string? packageId = (string?)reader.Value;
+                            reader.Read();
+                            IReadOnlyList<VulnerabilityInfo>? info = serializer.Deserialize<IReadOnlyList<VulnerabilityInfo>>(reader);
+
+                            if (packageId != null && info != null)
+                            {
+                                result.Add(packageId, info);
+                            }
+                        }
+                        break;
+
+                    default:
+                        throw new JsonSerializationException("Unexpected token type");
+                }
+            } while (!finished && reader.Read());
+
+            return new VulnerabilityFileData(result);
+        }
+
+        public override void WriteJson(JsonWriter writer, object? value, JsonSerializer serializer)
+        {
+            var data = (VulnerabilityFileData?)value;
+
+            if (data != null)
+            {
+                writer.WriteStartObject();
+
+                foreach (var kvp in data.OrderBy(kvp => kvp.Key, StringComparer.InvariantCultureIgnoreCase))
+                {
+#pragma warning disable CA1308 // Normalize strings to uppercase - NuGet has been using lower ids for years
+                    string lowerKey = kvp.Key.ToLowerInvariant();
```

Comment:
If the API endpoint only returned lowercase strings, this would not be needed.  Is that something worth considering?  I guess there'd be no guarantee that the server implementation abides by that...

Reply:
This is in the serialization code, not the deserialization code.  NuGet.Client will never use it, but other people might find it useful (maybe the server team in the future? it might be useful to other server implementors who want to generate vulnerability data for their implementation).

What I'm thinking is they use NuGet.Packaging to parse a bunch of nupkg files, whose nuspec files will use "package preferred" casing, not lowercase. They then populate this data structure and then use `string json = JsonConverter.Serialize(data)`. In order to comply with NuGet's protocol spec, it then needs to lowercase the package ids (and version strings, though that's only relevant for versions with pre-release strings, and it's also a violation of the SemVer spec, but that ship has long sailed).